src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -54,133 +54,60 @@ class GTPair(val p: Int) extends AnyVal {
 
-  require(_gt >= -1, s"invalid _gt value: ${ _gt }")
-  require(_dp >= -1, s"invalid _dp value: ${ _dp }")
+  val isDosage: Boolean
 
-  if (isDosage) {
-    require(_gq == -1)
-    if (_px == null)
-      require(_gt == -1)
-    else {
-      require(_px.sum == 32768)
-      require(_gt == Genotype.gtFromLinear(_px).getOrElse(-1))
-    }
-  }
+  override def equals(that: Any): Boolean
 
-  def check(nAlleles: Int) {
-    val nGenotypes = triangle(nAlleles)
-    assert(gt.forall(i => i >= 0 && i < nGenotypes))
-    assert(ad.forall(a => a.length == nAlleles))
-    assert(px.forall(a => a.length == nGenotypes))
-  }
+  override def hashCode: Int
 
-  def copy(gt: Option[Int] = this.gt,
-    ad: Option[Array[Int]] = this.ad,
-    dp: Option[Int] = this.dp,
-    gq: Option[Int] = this.gq,
-    px: Option[Array[Int]] = this.px,
-    fakeRef: Boolean = this.fakeRef,
-    isDosage: Boolean = this.isDosage): Genotype = Genotype(gt, ad, dp, gq, px, fakeRef, isDosage)
+  def unboxedGT: Int
 
-  override def equals(that: Any): Boolean = that match {
-    case g: Genotype =>
-      _gt == g._gt &&
-        ((_ad == null && g._ad == null)
-          || (_ad != null && g._ad != null && _ad.sameElements(g._ad))) &&
-        _dp == g._dp &&
-        _gq == g._gq &&
-        fakeRef == g.fakeRef &&
-        isDosage == g.isDosage &&
-        ((_px == null && g._px == null)
-          || (_px != null && g._px != null && _px.sameElements(g._px)))
+  def gt: Option[Int]
 
-    case _ => false
-  }
+  def ad: Option[Array[Int]]
 
-  override def hashCode: Int =
-    new HashCodeBuilder(43, 19)
-      .append(_gt)
-      .append(util.Arrays.hashCode(_ad))
-      .append(_dp)
-      .append(_gq)
-      .append(util.Arrays.hashCode(_px))
-      .append(fakeRef)
-      .append(isDosage)
-      .toHashCode
+  def dp: Option[Int]
 
-  def unboxedGT: Int = _gt
+  def od: Option[Int]
 
-  def gt: Option[Int] =
-    if (_gt >= 0)
-      Some(_gt)
-    else
-      None
+  def gq: Option[Int]
 
-  def ad: Option[Array[Int]] = Option(_ad)
+  def px: Option[Array[Int]]
 
-  def dp: Option[Int] =
-    if (_dp >= 0)
-      Some(_dp)
-    else
-      None
+  def pl: Option[Array[Int]]
 
-  def od: Option[Int] =
-    if (_dp >= 0 && _ad != null)
-      Some(_dp - _ad.sum)
-    else
-      None
+  def dosage: Option[Array[Double]]
 
-  def gq: Option[Int] =
-    if (_gq >= 0)
-      Some(_gq)
-    else
-      None
+  override def toString: String
 
-  def px: Option[Array[Int]] = Option(_px)
+  def toRow: Row
+
+  def toJSON: JValue
 
-  def pl: Option[Array[Int]] =
-    if (_px == null)
-      None
-    else if (!isDosage)
-      Some(_px)
-    else
-      Some(Genotype.linearToPhred(_px))
 
-  def dosage: Option[Array[Double]] =
-    if (_px == null)
-      None
-    else if (isDosage)
-      Some(_px.map(_ / 32768.0))
-    else
-      Some(Genotype.phredToDosage(_px))
 
-  def isHomRef: Boolean = Genotype.isHomRef(_gt)
+  def isHomRef: Boolean = Genotype.isHomRef(unboxedGT)
```

Comment:
I moved them all over, with unboxedGTs.  Seems like they were never called directly in the first place.